slick/src/main/scala/slick/util/AsyncExecutor.scala

```diff
@@ -56,19 +56,61 @@ object AsyncExecutor extends Logging {
 
     @volatile private[this] var executor: ThreadPoolExecutor = _
 
+    if (maxConnections > maxThreads) {
+      // NOTE: when using transactions or DB locks, it may happen that a task has a lock on the database but no thread
+      // to complete its action, while other tasks may have all the threads but are waiting for the first task to
+      // complete. This creates a deadlock.
+      logger.warn("Having maxConnection > maxThreads can result in deadlocks if transactions or database locks are used.")
+    }
+
     lazy val executionContext = {
       if(!state.compareAndSet(0, 1))
         throw new IllegalStateException("Cannot initialize ExecutionContext; AsyncExecutor already shut down")
       val queue: BlockingQueue[Runnable] = queueSize match {
-        case 0 => new SynchronousQueue[Runnable]
-        case -1 => new LinkedBlockingQueue[Runnable]
-        case n => new ManagedArrayBlockingQueue(maxConnections, n).asInstanceOf[BlockingQueue[Runnable]]
+        case 0 =>
+          // NOTE: SynchronousQueue does not schedule high-priority tasks before others and so it cannot be used when
+          // the number of connections is limited (lest high-priority tasks may be holding all connections and low/mid
+          // priority tasks all threads -- resulting in a deadlock).
+          require(maxConnections == Integer.MAX_VALUE, "When using queueSize == 0 (direct hand-off), maxConnections must be Integer.MAX_VALUE.")
+
+          new SynchronousQueue[Runnable]
+        case -1 =>
+          // NOTE: LinkedBlockingQueue does not schedule high-priority tasks before others and so it cannot be used when
+          // the number of connections is limited (lest high-priority tasks may be holding all connections and low/mid
+          // priority tasks all threads -- resulting in a deadlock).
+          require(maxConnections == Integer.MAX_VALUE, "When using queueSize == -1 (unlimited), maxConnections must be Integer.MAX_VALUE.")
+
+          new LinkedBlockingQueue[Runnable]
+        case n =>
+          // NOTE: The current implementation of ManagedArrayBlockingQueue is flawned. It makes the assumption that all
```

Comment:
I would say the exact opposite: The AsyncExecutor explicitly assumes that tasks either pass through the ManagedArrayBlockingQueue or don't pass through it. That's why there is this `inUseCounterSet` variable on the runnable:

- When the task is passed through the queue, the `increaseInUseCount` is called when it is taken from the queue and `inUseCounterSet` is set.
- When the task does not pass through the queue, it is set in the `beforeExecute` method.

In `inUseCounterSet` SHOULD protect us from increasing the in-use count twice.

Making the whole `increaseInUseCounter` method locked in #1854 , most likely fixed that.

Reply:
I still think my analysis is correct.

I saw the `beforeExecute` on the `ThreadPoolExecutor` which tries to do what you describe. 

However, it does not work since when the new thread is created, it could be that all connections are in use (`inUseCount == maximumInUse` in `ManagedArrayBlockingQueue`). Then the new thread will call `incrementInUseCount` anyway and this results in the `require(inUseCount < maximumInUse, "count cannot be increased")` failing.

Reply:
Ah yes, according to API the `ThreadPoolExecutor.beforeExecute` is allowed to throw an exception, but the task will then just not be executed, which is not desired.
Indeed, requiring `minThreads == maxThreads` and `allowCoreThreadTimeOut = false` (default) will make sure only that only the first `corePoolSize` created workers will bypass the queue and for these the `beforeExecute` will never throw if `minThreads` <= `maxConnection` of the ManagedArrayBlockingQueue.

It might be even better if we could remove the `AsyncExecutor.apply` factory method with all those extra parameters, introduced by https://github.com/slick/slick/commit/f3626f33e0c0ce0a0e7e453ede2a22f58125d971. 
It only causes problems, at least for the ManagedBlockingArrayQueue.
But maybe someone is using this in combination with the SynchronousQueue or the LinkedBlockingQueue?

Reply:
Haha: apparently my past self was already aware that `beforeExecute` can throw: https://github.com/slick/slick/pull/1461#issuecomment-205755838. I still think it's not that harmful.